### DISCUSSION TITLE
Optional Services

### DISCUSSION THREAD

**Discussion by atrauzzi** [microsoft/aspire#4527] [Ideas]
It would be interesting if Aspire supported some kind of "optional" flag for services.  Such a flag would be more geared towards the local development experience, ensuring that developers could keep resource on RAM-constrained dev setups usage to a minimum.

The Aspire dashboard (and Rider/VS/VS Code tooling) could feature a button to start/stop a service.  Including in debug mode.

**Comment by davidfowl** [comment]
The challenge with optional services will be dependencies that depend on values from them. I'm sure you're thinking about this from a "don't always start my project" POV, but it would be good to walk through the implications of on demand start and stop, especially if the dependency was never started in the first place.

A few starter questions:

1. Is this only for projects or is it any resource?
2. Projects need to be designed to expect the dependency to be optional right? That means if no connection information is provided (because the dependency is optional) the application doesn't blow up right?

**Reply by atrauzzi** [reply]
My initial thinking is that it wouldn't act on any knowledge of dependencies as part of this.  At its simplest, a project could be flagged as "don't auto start" in the config, and Aspire could blindly comply.

1. I feel like projects are more common, but the truth is people could also use this to trim down the startup time (read: inner loop!) of the overall profile
2. I was thinking Aspire would still provide all the "would be correct if it was running" connection information.  The only thing that the flag would impact is that the service isn't automatically started.
   - It would be an exercise to the dev to deal with the consequences
   - Devs could use the dashboard to tell Aspire to start the service, effectively bringing it back online at the connection information the other apps in the profile would have expected to see it at
   - Heck, this could even incentivize a better attitude and awareness towards degraded states and microservice design by making it a snap to simulate outages :bulb:  
4. Maybe this would merit some notion of sub-profiles or multiple profiles somehow selected by the run configuration? I'm not sold on this truthfully.  I would rather see the web UI have knobs to start and stop things

**Comment by runxc1** [comment]
So this is exactly what I am looking for.   My basic scenario is this.   I have a couple of APIs, a single sign on server and then 5-6 different apps that have common code.  Before looking at Aspire this was 5 different solutions but that has some downsides as common code could be affecting more than one of the apps.   Also the separate web apps are mostly related, client front end, administrator back end, client back end etc.   I'd like to be able to click on any of the 5 front ends and start them.  Anything they rely upon would also of course get started but most days I'm only working on one maybe 2 of them but still nice to see the wider picture.

**Reply by davidfowl** [reply]
A common ask, but not yet:
https://github.com/dotnet/aspire/issues/5851
https://github.com/dotnet/aspire/issues/303

**Comment by atrauzzi** [comment]
Should also consider implications of multi solution / git repo in all this.
